src/lib/flow/sol-flow-composed.c

```diff
@@ -0,0 +1,530 @@
+static int
+simple_port_process(struct sol_flow_node *node, void *data, uint16_t port,
+    uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct composed_node_data *cdata = data;
+    struct sol_flow_packet *composed;
+    uint16_t i;
+
+    if (cdata->inputs[port]) {
+        sol_flow_packet_del(cdata->inputs[port]);
+        cdata->inputs[port] = NULL;
+    }
+
+    cdata->inputs[port] = sol_flow_packet_dup(packet);
+    SOL_NULL_CHECK(cdata->inputs[port], -ENOMEM);
+
+    for (i = 0; i < cdata->inputs_len; i++) {
+        if (!cdata->inputs[i])
+            break;
+    }
+
+    if (i != cdata->inputs_len)
+        return 0;
+
+    composed = sol_flow_packet_new(cdata->composed_type, cdata->inputs);
+    SOL_NULL_CHECK(composed, -ENOMEM);
+    return sol_flow_send_packet(node, 0, composed);
+}
+
+static int
+composed_port_process(struct sol_flow_node *node, void *data, uint16_t port,
+    uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    int r;
+    uint16_t len, i;
+    struct sol_flow_packet **children, *out_packet;
+
+    r = sol_flow_packet_get_composed_members_len(
+        sol_flow_packet_get_type(packet), &len);
+    SOL_INT_CHECK(r, < 0, r);
+
+    children = alloca(len * sizeof(struct sol_flow_packet *));
+
+    r = sol_flow_packet_get(packet, children);
+    SOL_INT_CHECK(r, < 0, r);
+
+    for (i = 0; i < len; i++) {
+        out_packet = sol_flow_packet_dup(children[i]);
+        SOL_NULL_CHECK(out_packet, -ENOMEM);
+        r = sol_flow_send_packet(node, i, out_packet);
+        SOL_INT_CHECK(r, < 0, r);
+    }
+
+    return 0;
+}
+
+static int
+setup_simple_ports(struct sol_vector *in_ports, const struct sol_str_slice contents, bool is_input)
+{
+    struct sol_vector tokens;
+    struct sol_str_slice *slice;
+    struct composed_node_port_type *port_type;
+    const struct sol_flow_packet_type *packet_type;
+    struct sol_buffer buf;
+    char *token, *name, *type;
+    size_t i_slice;
+    uint16_t i;
+    int r;
+
+    sol_buffer_init(&buf);
+    for (i_slice = 0; i_slice < contents.len; i_slice++) {
+        if (isspace(contents.data[i_slice]))
+            continue;
+        sol_buffer_append_slice(&buf, SOL_STR_SLICE_STR(contents.data + i_slice, 1));
+    }
+
+    tokens = sol_util_str_split(sol_buffer_get_slice(&buf), DELIM, 0);
+
+    if (tokens.len < 2) {
+        SOL_ERR("A composed node must have at least two ports.");
+        sol_vector_clear(&tokens);
+        sol_buffer_fini(&buf);
+        return -EINVAL;
+    }
+
+    name = type = token = NULL;
+    SOL_VECTOR_FOREACH_IDX (&tokens, slice, i) {
+        token = strndup(slice->data, slice->len);
+        if (!token) {
+            r = -ENOMEM;
+            SOL_ERR("Could not alloc memory for the token");
+            goto err_exit;
+        }
+
+        if (sscanf(token, "%m[^(](%m[^)])", &name, &type) != 2) {
+            r = -errno;
+            SOL_ERR("Could not parse the arguments list");
+            goto err_exit;
+        }
```

Comment:
avoid doing the `strndup()` + `sscanf()` using %m for no good reason. In this fragment just use `memchr()` to find `'('` and `')'`. You can still duplicate `name`  as you need it later, but keep `type` as a slice.